### PR TITLE
Refactoring 3 uyuni

### PR DESCRIPTION
**Pull request by lucidd**
## What does this PR change?

- refactor SaltReactor to avoid `setReactor` on `SystemQuery`
- Create initial `SaltApi` interface and move some methods there.
- Add VirtManager interface and move virtualization methods from `SystemQuery` there.
- reduce uses of `SaltService.INSTANCE` further.

- [x] **DONE**

## Documentation
- No documentation needed: no user visible change

- [x] **DONE**

## Test coverage
- No tests: already covered

- [x] **DONE**

## Links

Fixes #
Tracks # **add downstream PR, if any**

- [ ] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [x] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"		 
- [ ] Re-run test "java_pgsql_tests"		 
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"		 
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"		 
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/suse/manager/webui/services/iface/RedhatProductInfo.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+     * @param whatProvidesResIn what provides res result
+     */
+    public RedhatProductInfo(Optional<String> centosReleaseContentIn, Optional<String> rhelReleaseContentIn,
+                             Optional<String> whatProvidesResIn) {
+        this.centosReleaseContent = centosReleaseContentIn;
+        this.rhelReleaseContent = rhelReleaseContentIn;
+        this.whatProvidesRes = whatProvidesResIn;
+    }
+
+
```
Comment (renner) [inline java/code/src/com/suse/manager/webui/services/iface/RedhatProductInfo.java:40]:
```suggestion
```

File: java/code/src/com/suse/manager/webui/services/iface/RedhatProductInfo.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+ * implied, including the implied warranties of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. You should have received a copy of GPLv2
+ * along with this software; if not, see
+ * http://www.gnu.org/licenses/old-licenses/gpl-2.0.txt.
+ *
+ * Red Hat trademarks are not licensed under GPLv2. No permission is
+ * granted to use or replicate Red Hat trademarks that are incorporated
+ * in this software or its documentation.
+ */
+package com.suse.manager.webui.services.iface;
```
Comment (renner) [inline java/code/src/com/suse/manager/webui/services/iface/RedhatProductInfo.java:15]: Is `iface` the right package to use here?

**Comment by lucidd** [comment]
> Looks good to me, but the copyright dates of those changed files should be changed to a range finishing by 2020. And ideally where there are two lines of copyright I'ld move the SUSE one to the first one since it's the shown line when the comment is folded in IDEs

This should really be done by an automated process otherwise this will always be out of sync. There are many files with outdated copyright header that have been changed many times since then.

**Comment by cbosdo** [comment]
> > Looks good to me, but the copyright dates of those changed files should be changed to a range finishing by 2020. And ideally where there are two lines of copyright I'ld move the SUSE one to the first one since it's the shown line when the comment is folded in IDEs
> 
> This should really be done by an automated process otherwise this will always be out of sync. There are many files with outdated copyright header that have been changed many times since then.

For sure an automated tool would be good for that.

File: java/code/src/com/suse/manager/webui/services/iface/RedhatProductInfo.java (hunk 2 above)
Comment (lucidd) [reply java/code/src/com/suse/manager/webui/services/iface/RedhatProductInfo.java:15]: do we have a better place? i put it there so its close to the interface its used in.

**Comment by lucidd** [comment]
> > > Looks good to me, but the copyright dates of those changed files should be changed to a range finishing by 2020. And ideally where there are two lines of copyright I'ld move the SUSE one to the first one since it's the shown line when the comment is folded in IDEs
> > 
> > 
> > This should really be done by an automated process otherwise this will always be out of sync. There are many files with outdated copyright header that have been changed many times since then.
> 
> For sure an automated tool would be good for that.

yes but that's a topic for another PR not this one so if everything else is ok could you approve it?

**Comment by cbosdo** [comment]
> > For sure an automated tool would be good for that.
> 
> yes but that's a topic for another PR not this one so if everything else is ok could you approve it?

Nope, I'ld still like to see the copyright lines that are in the diff of this PR to be OK even if you don't set all the other ones right

File: java/code/src/com/suse/manager/webui/services/impl/SaltService.java [hunk 3]
```diff
@@ -122,14 +123,16 @@
...
 
     private final Batch defaultBatch;
 
     /**
      * Singleton instance of this class
      */
-    public static final SystemQuery INSTANCE = new SaltService();
+    private static final SaltService INSTANCE_SALT_SERVICE = new SaltService();
+    public static final SystemQuery INSTANCE = INSTANCE_SALT_SERVICE;
+    public static final SaltApi INSTANCE_SALT_API = INSTANCE_SALT_SERVICE;
```
Comment (admd) [inline java/code/src/com/suse/manager/webui/services/impl/SaltService.java:135]: This is confusing, why they both are pointing to same? Are they even being used?

Comment (lucidd) [reply java/code/src/com/suse/manager/webui/services/impl/SaltService.java:135]: The reason is to have the variable be declared with the interface type rather then `SaltService` to there is no accidental mixup. It also helps to track where each interface is used right now. 
In the end those static instances are just temporary until further refactorings.

**Pull request merged**